### PR TITLE
Able to build the installer in Visual Studio 2017 MSBuild15 environment

### PR DESCRIPTION
**Pull request by EbenZhang**
Changes proposed in this pull request:
 - Able to build the installer using msbuild15/vs2017

How did I test this code:
 - Run the buildinstaller.vs2017.cmd in the setup folder. 
 - Install the complete_setup with the explore integration ticked
 - Right click a repo and verify the gitextensions menu items are still working.
 - Push a branch to a repo of a new host e.g. bitbucket, make sure the openssh will ask for password

i'm not able to test the buildinstaller.vs2015.cmd, can someone verify that buildinstaller.vs2015.cmd is still working

Has been tested on (remove any that don't apply):
 - GIT 2.10 and above
 - Windows 10 64-bit

### CODE REVIEW

**Comment by jbialobr** [comment]
> i'm not able to test the buildinstaller.vs2015.cmd, can someone verify that buildinstaller.vs2015.cmd is still working

AppVeyor builds using that script.

Can you move vswhere.ext to the `tools` dir?

**Comment by KindDragon** [comment]
I think better download it using curl in like in `DownloadExternals.cmd`

**Comment by EbenZhang** [comment]
@KindDragon  

`DownloadExternals.cmd` will be called by the `makeinstaller.cmd` after the projects are built. Do you think we should move it to buildinstaller.vs201x.cmd? Another option is to NOT touch the `downloadExternals.cmd` but download the `vswhere` in `buildinstaller.vs2017.cmd`.

**Comment by KindDragon** [comment]
> Another option is to NOT touch the downloadExternals.cmd but download the vswhere in buildinstaller.vs2017.cmd.

I suggest using this option

**Comment by vbjay** [comment]
Downloading the CMD file would put it in a downloaded file security level.
Would that affect being able to run it?

**Comment by EbenZhang** [comment]
> Downloading the CMD file would put it in a downloaded file security level.
> Would that affect being able to run it?

The file is an exe not CMD.  
Not sure what is the downloaded file security level, it runs OK probably because the file is signed by Microsoft.

**Comment by EbenZhang** [comment]
squashed the commits.

**Comment by jbialobr** [comment]
The vswhere.exe should be cached on AppVeyor
https://github.com/gitextensions/gitextensions/commit/99dc362f1038becf01740f8169b425ef99db7c08#diff-180360612c6b8c4ed830919bbb4dd459R15

File: Setup/BuildInstallers.VS2017.cmd [hunk 1]
```diff
@@ -0,0 +1,49 @@
+@echo off
+
+cd /d "%~p0"
+
+IF NOT EXIST "%~p0\tools\vswhere1.0.62.exe" (
+    "%~p0\tools\curl.exe" -L -k -o %~p0\tools\vswhere1.0.62.exe https://github.com/Microsoft/vswhere/releases/download/1.0.62/vswhere.exe -L https://github.com > NUL
```
Comment (KindDragon) [inline Setup/BuildInstallers.VS2017.cmd]: Maybe use name `vswhere.exe`? So that when we update it, the old version does not remain on the disk

Comment (EbenZhang) [inline Setup/BuildInstallers.VS2017.cmd]: Is there anyway to know the local vswhere.exe is out of date without the version in file name?

**Comment by EbenZhang** [comment]
> The vswhere.exe should be cached on AppVeyor
> 99dc362#diff-180360612c6b8c4ed830919bbb4dd459R15

Should I add an entry of ` - Setup/tools` in the cache section? How can we invalid the cache if we have a newer version?

**Comment by KindDragon** [comment]
> Should I add an entry of - Setup/tools in the cache section?

AppVeyor support only caching folders 😞 It will cache our committed files

> How can we invalid the cache if we have a newer version?

You can set dependencies 
https://www.appveyor.com/docs/build-cache/

You can move tool to `cache` (sorry about another moving :) ) and dependency:
```
cache: 
  - Setup\cache -> Setup\DownloadExternals.cmd, Setup\BuildInstallers.VS2017.cmd
```

**Comment by KindDragon** [comment]
https://blogs.msdn.microsoft.com/heaths/2017/04/21/vswhere-is-now-installed-with-visual-studio-2017/

**Comment by EbenZhang** [comment]
@KindDragon updated the script to use vswhere shipped with visual studio installer

File: Setup/BuildInstallers.VS2017.cmd [hunk 2]
```diff
@@ -2,12 +2,19 @@
...
+    set vswhere="%ProgramFiles%\%subPathToVsWhere%"
 )
 
-for /f "usebackq tokens=1* delims=: " %%i in (`.\tools\vswhere1.0.62.exe -latest -requires Microsoft.Component.MSBuild`) do (
+if not exist %vswhere% (
+    echo "Failed to find vswhere.exe, make sure you have installed Visual Studio 15.2.26418.1 or a later version."
+    exit /B 1
+)
+
+for /f "usebackq tokens=1* delims=: " %%i in (`%vswhere% -latest -requires Microsoft.Component.MSBuild`) do (
```
Comment (KindDragon) [inline Setup/BuildInstallers.VS2017.cmd]: Maybe add also `-version "[15.0,16.0)"`?

File: .gitignore [hunk 3]
```diff
@@ -66,3 +66,4 @@ UpgradeLog*.htm*
 packages/
 GitExtensions.*.sln.VisualState.xml
 GitExtensions.settings.backup
+vswhere*.exe
```
Comment (KindDragon) [inline .gitignore]: revert that

**Comment by EbenZhang** [comment]
@KindDragon reverted.

**Pull request merged**